File: test/Coypu.Tests/When_making_browser_interactions_robust/When_I_want_to_make_direct_calls_to_the_robust_wrapper.cs
using System;
using System.Linq;
using Coypu.Tests.TestDoubles;
using Coypu.Tests.When_interacting_with_the_browser;
using NUnit.Framework;

package Coypu.Tests.When_making_browser_interactions_robust
{
    [TestFixture]
    public class When_I_want_to_make_direct_calls_to_the_robust_wrapper : BrowserInteractionTests
    {
        @Test
        public void RobustAction_is_exposed_on_the_session()
        {
            var calledOnWrapper = false;
            browserSession.RetryUntilTimeout(() =>
            {
                calledOnWrapper = true;
            });
            spyRobustWrapper.QueriesRan<object>().First().Run();
            assertThat(calledOnWrapper, Is.True);
        }

        @Test
        public void RobustFunction_is_exposed_on_the_session()
        {
            Func<string> function = () => "The expected result";

            spyRobustWrapper.StubQueryResult(SpyRobustWrapper.NO_EXPECTED_RESULT, "immediate result");

            assertThat(browserSession.RetryUntilTimeout(function), Is.EqualTo("immediate result"));

            var query = spyRobustWrapper.QueriesRan<string>().First();
            query.Run();

            assertThat(query.Result, Is.EqualTo("The expected result"));
        }

        @Test
        public void TryUntil_is_exposed_on_the_session()
        {
            var tried = false;
            var triedUntil = false;
            Action tryThis = () => tried = true;
            Func<bool> until = () => triedUntil = true;
            var overallTimeout = TimeSpan.FromMilliseconds(1234);

            var options = new Options { Timeout = overallTimeout };
            browserSession.TryUntil(tryThis, until,TimeSpan.Zero,options);

            var tryUntil = spyRobustWrapper.DeferredTryUntils[0];

            assertThat(tried, Is.False);
            tryUntil.TryThisBrowserAction.Act();
            assertThat(tried, Is.True);

            assertThat(triedUntil, Is.False);
            tryUntil.Until.Run();
            assertThat(triedUntil, Is.True);

            assertThat(tryUntil.OverallTimeout, Is.EqualTo(overallTimeout));
        }

        @Test
        public void Query_is_exposed_on_the_session()
        {
            Func<string> query = () => "query result";

            spyRobustWrapper.StubQueryResult("expected query result", "immediate query result");

            assertThat(browserSession.Query(query, "expected query result"), Is.EqualTo("immediate query result"));

            var robustQuery = spyRobustWrapper.QueriesRan<string>().First();
            robustQuery.Run();

            assertThat(robustQuery.Result, Is.EqualTo("query result"));
        }
    }
}

File: src/test/java/Coypu.Tests/When_making_browser_interactions_robust/When_I_want_to_make_direct_calls_to_the_robust_wrapper.java
package Coypu.Tests.When_making_browser_interactions_robust;

import Coypu.Actions.BrowserAction;
import Coypu.Options;
import Coypu.Queries.PredicateQuery;
import Coypu.Queries.Query;
import Coypu.Tests.TestDoubles.SpyRobustWrapper;
import Coypu.Tests.When_interacting_with_the_browser.BrowserInteractionTests;
import Coypu.TimeSpan;
import org.junit.Test;

import static org.hamcrest.core.Is.is;
import static org.hamcrest.core.IsEqual.equalTo;
import static org.junit.Assert.assertThat;

public class When_I_want_to_make_direct_calls_to_the_robust_wrapper extends BrowserInteractionTests
{
    @Test
    public void retryUntilTimeout_is_exposed_on_the_session()
    {
        final boolean[] calledOnWrapper = {false};
        browserSession.retryUntilTimeout(new BrowserAction(configuration) {
            public void act() {
                calledOnWrapper[0] = true;
            }
        });
        spyRobustWrapper.queriesRan().get(0).run();
        assertThat(calledOnWrapper[0], is(true));
    }

    @Test
    public void query_is_exposed_on_the_session()
    {
        Query<String> function = new Query<String>() {
            @Override
            public String run() {
                return "The expected result";
            }

            @Override
            public String getExpectedResult() {
                return null;
            }
        };

        spyRobustWrapper.stubQueryResult(SpyRobustWrapper.NO_EXPECTED_RESULT, "immediate result");

        assertThat(browserSession.query(function), is(equalTo("immediate result")));

        Query<String> query = spyRobustWrapper.queriesRan().get(0);
        String result = query.run();

        assertThat(result, is(equalTo("The expected result")));
    }

    @Test
    public void tryUntil_is_exposed_on_the_session()
    {
        final boolean[] tried = {false};
        final boolean[] triedUntil = {false};
        BrowserAction tryThis = new BrowserAction() {
            @Override
            public void act() {
                tried[0] = true;
            }
        };
        
        PredicateQuery until = new PredicateQuery() {
            @Override
            public Boolean predicate() {
                triedUntil[0] = true;
                return true;
            }
        };
        TimeSpan overallTimeout = TimeSpan.fromMilliseconds(1234);

        Options options = new Options();
        options.Timeout = overallTimeout;
        browserSession.tryUntil(tryThis, until,TimeSpan.zero(),options);

        SpyRobustWrapper.TryUntilArgs tryUntil = spyRobustWrapper.DeferredTryUntils.get(0);

        assertThat(tried[0], is(false));
        tryUntil.TryThisBrowserAction.act();
        assertThat(tried[0], is(true));

        assertThat(triedUntil[0], is(false));
        tryUntil.Until.run();
        assertThat(triedUntil[0], is(true));

        assertThat(tryUntil.OverallTimeout, is(equalTo(overallTimeout)));
    }
}